pallets/deitos/src/tests.rs

```diff
@@ -1,2 +1,223 @@
-use crate::{mock::*, Error, Event};
+use super::*;
+use crate::{mock::*, types::*, Error, Event};
 use frame_support::{assert_noop, assert_ok};
+
+fn register_ip() {
+    let total_storage: Storage = 10000000_u32.into();
+    assert_ok!(Deitos::register_ip(RuntimeOrigin::signed(1), total_storage));
+}
+
+#[test]
+fn test_correct_ip_registration() {
+    new_test_ext().execute_with(|| {
+        let total_storage: Storage = 10000000_u32.into();
+        let price_storage_per_block: BalanceOf<Test> = 1000_u64.into();
+
+        register_ip();
+
+        assert_eq!(
+            <Balances as fungible::InspectHold<_>>::balance_on_hold(
+                &HoldReason::IPInitialDeposit.into(),
+                &1
+            ),
+            DEPOSIT_AMOUNT
+        );
+
+        assert_eq!(
+            <Balances as fungible::Inspect<_>>::balance(&1),
+            INITIAL_BALANCE - DEPOSIT_AMOUNT
+        );
+
+        assert_eq!(
+            InfrastructureProvider::<Test>::get(1).unwrap(),
+            IPDetails::<Test> {
+                total_storage,
+                reserved_storage: Zero::zero(),
+                status: IPStatus::Validating,
+                active_agreements: BoundedVec::new(),
+                deposit_amount: DEPOSIT_AMOUNT,
+            }
+        );
+
+        System::assert_has_event(RuntimeEvent::Deitos(pallet_deitos::Event::IPRegistered {
+            ip: 1,
+            total_storage,
+        }));
+    });
+}
+
+#[test]
+fn test_fail_ip_registration_already_exists() {
+    new_test_ext().execute_with(|| {
+        let ip = 1;
+        // Register the IP once
+        register_ip();
+
+        // Attempt to register the same IP again and expect failure
+        let total_storage: Storage = 10000000_u32.into();
+        assert_noop!(
+            Deitos::register_ip(RuntimeOrigin::signed(ip), total_storage),
+            Error::<Test>::IPAlreadyExists
+        );
+    });
+}
+
+#[test]
+fn test_update_storage_cost_per_unit() {
```

Comment:
Does it make sense to add a test verifying that after a storage cost update the deposit is released properly during an IP unregister?

Reply:
hmmm i'm not able to see the reason why that unrelease should happen during a storage update call. I do believe that check should happen for unregistering which is already covered in the test: `test_unregister_ip`

Reply:
`test_unregister_ip`
1. registers ip with deposit
2. unregisters ip
3.  checks that deposit is returned

what I mean is a little different scenarion:
1. register ip with `deposit1`
2. change `deposit1` to `deposit2`
3. unregister ip
4. check that `deposit1` is returned (not `deposit2`)

wyt

Reply:
still not getting it ... specifically because of this:

`change deposit1 to deposit2`

Deposits will be static and it will happen only once during registration. I don't see how deposits should be changed with storage upadtes. Maybe i'm missing something?

Reply:
I think you can update it with a runtime upgrade. But for the test you can just change the value in `IPDepositAmount`. The test will just make sure that the released amount is invariant to `IPDepositAmount`.

Reply:
ah ok .... the deposit_amount for each IP is now get stored in the IP storage item... so it should work like such. 

good point, i will implement it soon.

Reply:
yes, I don't see any problems with your implementation. just suggesting this test for future regressions.